internal/cuegen/component.go

```diff
@@ -0,0 +1,92 @@
+package cuegen
+
+import (
+	"fmt"
+
+	"cuelang.org/go/cue/ast"
+	"cuelang.org/go/cue/token"
+	"github.com/benthosdev/benthos/v4/internal/docs"
+)
+
+type componentOptions struct {
+	collectionIdent  *ast.Ident
+	disjunctionIdent *ast.Ident
+
+	canLabel      bool
+	canPreProcess bool
+}
+
+func doComponents(specs []docs.ComponentSpec, opts *componentOptions) ([]ast.Decl, error) {
+	fields := make([]interface{}, 0, len(specs))
+
+	for _, v := range specs {
+		field, err := doComponentSpec(v)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate cue type for component: %s: %w", v.Name, err)
+		}
+		fields = append(fields, field)
+	}
+
+	var addons []interface{}
+	if opts.canLabel {
+		addons = append(
+			addons,
+			ast.NewIdent("label"),
+			token.OPTION,
+			ast.NewIdent("string"),
+		)
+	}
+
+	if opts.canPreProcess {
+		addons = append(
+			addons,
+			ast.NewIdent("processors"),
+			token.OPTION,
+			ast.NewList(&ast.Ellipsis{Type: ast.NewIdent("#Processor")}),
+		)
+	}
+
+	decls := []ast.Decl{
+		&ast.Field{
+			Label: opts.collectionIdent,
+			Value: ast.NewStruct(fields...),
+		},
+		&ast.Field{
+			Label: opts.disjunctionIdent,
+			Value: ast.NewCall(ast.NewIdent("or"), ast.NewList(&ast.Comprehension{
+				Clauses: []ast.Clause{
+					&ast.ForClause{
+						Key:    ast.NewIdent("name"),
+						Value:  ast.NewIdent("config"),
+						Source: opts.collectionIdent,
+					},
+				},
+				Value: ast.NewStruct(&ast.Field{
+					Label: interpolateIdent(ast.NewIdent("name")),
+					Value: ast.NewIdent("config"),
+				}),
+			})),
+		},
+	}
+
+	if len(addons) > 0 {
+		decls = append(decls, &ast.Field{
+			Label: opts.disjunctionIdent,
+			Value: ast.NewBinExpr(
+				token.AND,
+				opts.disjunctionIdent,
+				ast.NewStruct(
```

Comment:
if you've got some `*ast.Field` (let's call it `myfield`) that you want to include in a Cue struct then you can go right ahead and do it with `ast.NewStruct(myfield)`. HOWEVER, if that field should be marked as optional then you need something a little more tricky like so: `ast.NewStruct(myfield.Label, token.Option, myField.Value)`. You will see this pattern in a few places in this PR.